File: tyto/go/core/typeutil/uint.go
package typeutil

// 用high和low组成一个新的uint，high放高位，low放低位
func MakeUint16(high uint8, low uint8) uint16 {
	data := uint16(high) << 8
	data |= uint16(low)
	return data
}

func MakeUint32(high uint16, low uint16) uint32 {
	data := uint32(high) << 16
	data |= uint32(low)
	return data
}

func MakeUint64(high uint32, low uint32) uint64 {
	data := uint64(high) << 32
	data |= uint64(low)
	return data
}

// 获取高位部分
func HighPartUint16(data uint16) uint8 {
	data >>= 8
	return uint8(data)
}

func HighPartUint32(data uint32) uint16 {
	data >>= 16
	return uint16(data)
}

func HighPartUint64(data uint64) uint32 {
	data >>= 32
	return uint32(data)
}

// 获取低位部分
func LowPartUint16(data uint16) uint8 {
	data &= 0xff
	return uint8(data)
}

func LowPartUint32(data uint32) uint16 {
	data &= 0xffff
	return uint16(data)
}

func LowPartUint64(data uint64) uint32 {
	data &= 0xffffffff
	return uint32(data)
}

File: tyto/csharp/core/Tyto.Utils/TypeUtil/Uint.cs
namespace Tyto.Utils;

using uint8 = byte;
using uint16 = ushort;
using uint32 = uint;
using uint64 = ulong;

/// <summary>
/// 整数合并和分拆处理类
/// </summary>
public static partial class TypeUtil
{
	/// <summary>
	/// 将high和low组合为一个新的整数，high放高位部分，low放低位部分
	/// </summary>
	public static uint16 MakeUint16(uint8 high, uint8 low)
	{
		uint16 h = high;
		uint16 l = low;
		return (uint16)((h << 8) | l);
	}

	/// <inheritdoc cref="MakeUint16"/>
	public static uint32 MakeUint32(uint16 high, uint16 low)
	{
		uint32 h = high;
		uint32 l = low;
		return (h << 16) | l;
	}

	/// <inheritdoc cref="MakeUint16"/>
	public static uint64 MakeUint64(uint32 high, uint32 low)
	{
		uint64 h = high;
		uint64 l = low;
		return (h << 32) | l;
	}

	/// <summary>
	/// 获取整数的高位部分
	/// </summary>
	public static uint8 HighPartUint16(uint16 val)
	{
		return (uint8)(val >> 8);
	}

	/// <inheritdoc cref="HighPartUint16"/>
	public static uint16 HighPartUint32(uint32 val)
	{
		return (uint16)(val >> 16);
	}

	/// <inheritdoc cref="HighPartUint16"/>
	public static uint32 HighPartUint64(uint64 val)
	{
		return (uint32)(val >> 32);
	}

	/// <summary>
	/// 获取整数的低位部分
	/// </summary>
	public static uint8 LowPartUint16(uint16 val)
	{
		return (uint8)(val & 0xFF);
	}

	/// <inheritdoc cref="LowPartUint16"/>
	public static uint16 LowPartUint32(uint32 val)
	{
		return (uint16)(val & 0xFFFF);
	}

	/// <inheritdoc cref="LowPartUint16"/>
	public static uint32 LowPartUint64(uint64 val)
	{
		return (uint32)(val & 0xFFFFffff);
	}
}